src/vs/workbench/contrib/terminal/node/terminalProfiles.ts

```diff
@@ -0,0 +1,154 @@
+import { ITerminalProfile } from 'vs/workbench/contrib/terminal/common/terminal';
+import * as cp from 'child_process';
+
+export interface IStatProvider {
+	stat(path: string): boolean,
+	lstat(path: string): boolean
+}
+
+export function detectAvailableShells(detectWslShells?: boolean, statProvider?: IStatProvider): Promise<ITerminalProfile[]> {
+	return platform.isWindows ? detectAvailableWindowsShells(detectWslShells, statProvider) : detectAvailableUnixShells();
+}
+
+async function detectAvailableWindowsShells(detectWslShells?: boolean, statProvider?: IStatProvider): Promise<ITerminalProfile[]> {
+	// Determine the correct System32 path. We want to point to Sysnative
+	// when the 32-bit version of VS Code is running on a 64-bit machine.
+	// The reason for this is because PowerShell's important PSReadline
+	// module doesn't work if this is not the case. See #27915.
+	const is32ProcessOn64Windows = process.env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
+	const system32Path = `${process.env['windir']}\\${is32ProcessOn64Windows ? 'Sysnative' : 'System32'}`;
+
+	let useWSLexe = false;
+
+	if (getWindowsBuildNumber() >= 16299) {
+		useWSLexe = true;
+	}
+
+	let expectedLocations: { [key: string]: string[] } = {
+		'Command Prompt': [`${system32Path}\\cmd.exe`],
+		'WSL Bash': [`${system32Path}\\${useWSLexe ? 'wsl.exe' : 'bash.exe'}`],
+		'Git Bash': [
+			`${process.env['ProgramW6432']}\\Git\\bin\\bash.exe`,
+			`${process.env['ProgramW6432']}\\Git\\usr\\bin\\bash.exe`,
+			`${process.env['ProgramFiles']}\\Git\\bin\\bash.exe`,
+			`${process.env['ProgramFiles']}\\Git\\usr\\bin\\bash.exe`,
+			`${process.env['LocalAppData']}\\Programs\\Git\\bin\\bash.exe`,
+		],
+		'Cygwin': [
+			`${process.env['HOMEDRIVE']}\\cygwin64\\bin\\bash.exe`,
+			`${process.env['HOMEDRIVE']}\\cygwin\\bin\\bash.exe`
+		]
+	};
+
+	// Add all of the different kinds of PowerShells
+	for await (const pwshExe of enumeratePowerShellInstallations()) {
+		expectedLocations[pwshExe.displayName] = [pwshExe.exePath];
+	}
+	const promises: Promise<ITerminalProfile | undefined>[] = [];
+	Object.keys(expectedLocations).forEach(key => promises.push(validateShellPaths(key, expectedLocations[key], statProvider)));
+	const shells = await Promise.all(promises);
+	let output = '';
+	cp.exec('wsl.exe -l', (err, stdout) => {
+		output = stdout;
+	});
+
+	let wslShell = shells.find(shell => shell?.path.endsWith('wsl.exe'));
+	if (detectWslShells && wslShell && output && output.length > 0) {
+		output.split('\n').forEach(distro => {
+			wslShell!.profileName += ` ${distro}`;
+			wslShell!.args = `-d ${distro}`;
+			shells.push(wslShell);
+		});
+	}
+
+	let cygwin = shells.find(shell => shell?.profileName === 'Cygwin');
+	let bashZsh = shells.find(shell => shell?.profileName === 'Git Bash' || shell?.path.endsWith('zsh.exe'));
+
+	if (cygwin) {
+		cygwin.args = ['-l'];
+	}
+
+	if (bashZsh) {
+		bashZsh.args = ['--login'];
+	}
```

Comment:
Consider merging this into the expectedLocations and maybe call it `autoDiscoveredShell`? That way the path and args are in the same spot and we don't need to iterate over `shells` multiple times afterwards.